### PR TITLE
[Proposal] Builder generation for CreateCollection and VectorParams

### PR DESCRIPTION
**Pull request by JojiiOfficial**
Proposal to for a builder pattern based `CreateCollection`. For this [derive_builder](https://docs.rs/derive_builder/latest/derive_builder/) is used.

A simple example for creating a collection:
```rust
CreateCollectionBuilder::default()
  .collection_name("name") // <- types implementing Into<String> are supported (String, &String, &str, ...)
  .vectors_config(
      VectorParamsBuilder::default()
          .distance(Distance::Cosine)
          .size(10),
          .on_disk(true)
  )
  .build()
```

And for example with a custom Hnsw config it could look like this:
```rust
CreateCollectionBuilder::default()
  .collection_name(collection_name)
  .vectors_config(
      VectorParamsBuilder::default()
          .distance(Distance::Cosine)
          .hnsw_config(
              HnswConfigDiffBuilder::default()
                  .m(12)
                  .on_disk(true)
                  .payload_m(13),
          )
          .size(10),
  )
  .build()
```

### CODE REVIEW

**Review by generall — APPROVED** [review]
From the user perspective - LGTM,
but I would like to request a review from @timvisee, as I don't really have experience in building macro

File: examples/search.rs [hunk 1]
```diff
@@ -27,6 +27,7 @@ async fn main() -> Result<()> {
     let collection_name = "test";
     client.delete_collection(collection_name).await?;
 
+    /*
```
Comment (timvisee) [inline examples/search.rs:30]: Let's just remove the old example here.

File: examples/search.rs [hunk 2]
```diff
@@ -40,6 +41,20 @@ async fn main() -> Result<()> {
             ..Default::default()
         })
         .await?;
+     */
+
+    client
+        .create_collection(
+            &CreateCollectionBuilder::default()
```
Comment (timvisee) [inline examples/search.rs:48]: I see we take a reference here. That is just a side effect of our function signature here, isn't it?

We _can_ change them into `AsRef<CollectionConfig>` to also allow passing in owned types. But lets leave it now and explore that when we rewrite our functions with the builder pattern.

File: tests/protos.rs [hunk 3]
```diff
@@ -1,19 +1,213 @@
...
+    append_file_to_file(GRPC_OUTPUT_FILE, "./tests/protos_append/grpc_macros.rs");
+    add_builder_macro_impls(GRPC_OUTPUT_FILE, builder_derive_options());
+
+    panic!("proto definitions changed. Stubs recompiled. Please commit the changes.")
+}
+
+/// Derive options for structs. (Path, build attributes, 'from' macro generation enabled)
+type BuildDeriveOptions = (&'static str, &'static str, bool);
+
+/// Extension to [`Builder`] to configure validation attributes.
```
Comment (timvisee) [inline tests/protos.rs:38]:
```suggestion
/// Extension to [`Builder`] to configure builder attributes.
```

**Review by timvisee — APPROVED** [review]
Cool!

@generall I used derive builder quite a bit before, which is why I suggested it to @JojiiOfficial. I think it'll be the right way forward.

File: examples/search.rs (hunk 2 above)
Comment (JojiiOfficial) [reply examples/search.rs:48]: If we're introducing breaking changes anyways, we could also take `Into<CreateCollection>` here and add implementations for `From<&mut CreateCollectionBulider>` and `From<&CreateCollection>` for CreateCollection so we don't even have to call .build(). But this is just sugar at this point.

**Pull request merged**